### PR TITLE
Feature/migrations models

### PR DESCRIPTION
**Pull request by zhamza372**

### CODE REVIEW

File: db/migrate/20200907135535_create_users.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
+      t.string :first_name, null:false
```
Comment (bk-az) [inline db/migrate/20200907135535_create_users.rb]: space `null:false` -> `null: false`

File: db/migrate/20200907135535_create_users.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :last_name, null:false
+      t.string :email, null:false, uniqueness:true
+      t.integer :role_id , null:false
+      t.integer :company_id ,null:false
+      t.integer :sequence_number , null:false
+      t.integer :department_id , null:false
```
Comment (bk-az) [inline db/migrate/20200907135535_create_users.rb]: department_id can be null

File: db/migrate/20200907135535_create_users.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+      t.boolean :active
```
Comment (bk-az) [inline db/migrate/20200907135535_create_users.rb]: null false, default value true

File: db/migrate/20200907135832_create_teams.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class CreateTeams < ActiveRecord::Migration[6.0]
+  def change
+    create_table :teams do |t|
+
+			t.string :name, null:false
```
Comment (bk-az) [inline db/migrate/20200907135832_create_teams.rb]: indentation

File: db/migrate/20200907135832_create_teams.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+			t.string :description
```
Comment (bk-az) [inline db/migrate/20200907135832_create_teams.rb]: description should text type

File: db/migrate/20200907135905_create_tasks.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+class CreateTasks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :tasks do |t|
+      t.integer :company_id , null:false
+      t.string :title , null:false
+      t.string :description
+      t.integer :assigned_to_id , null:false
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: can be null

File: db/migrate/20200907135905_create_tasks.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+      t.integer :assigned_to_type , null:false
+      t.integer :created_by_id , null:false
+      t.string :status , default: "Assigned"
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: default New, should be integer, use mappings

File: db/migrate/20200907135905_create_tasks.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+      t.datetime :due_date
+      t.integer :project_id , null:false
+      t.integer :reviewer_id 
+      t.integer :watcher_id
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: watcher_type

File: db/migrate/20200908070038_create_attendances.rb [hunk 9]
```diff
@@ -0,0 +1,12 @@
+class CreateAttendances < ActiveRecord::Migration[6.0]
+  def change
+    create_table :attendances do |t|
+    	t.integer :company_id , null:false
+			t.integer :user_id , null:false
+			t.datetime :login_time
+			t.datetime :logout_time
+			t.boolean :present
```
Comment (bk-az) [inline db/migrate/20200908070038_create_attendances.rb]: default true

File: db/migrate/20200908070148_create_comments.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[6.0]
+  def change
+    create_table :comments do |t|
+    	t.integer :company_id , null:false
+    	t.integer :commentable_id , null:false
+    	t.integer :commentable_type , null:false
+    	t.string :content
```
Comment (bk-az) [inline db/migrate/20200908070148_create_comments.rb]: text

File: db/migrate/20200908070243_create_time_logs.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class CreateTimeLogs < ActiveRecord::Migration[6.0]
+  def change
+    create_table :time_logs do |t|
+    	t.integer :task_id , null:false
+    	t.integer :company_id , null:false
+    	t.integer :user_id , null:false
+    	t.decimal :hours, :precision => 5, :scale => 3
```
Comment (bk-az) [inline db/migrate/20200908070243_create_time_logs.rb]: use new hash syntax

File: app/models/attachment.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+class Attachment < ApplicationRecord
```
Comment (bk-az) [inline app/models/attachment.rb:4]: define associations in all models

File: config/database.yml [hunk 13]
```diff
@@ -14,7 +14,7 @@ default: &default
   encoding: utf8mb4
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   username: root
-  password:
+  password: root
```
Comment (abdul674) [inline config/database.yml]: passwords should never be added in the repo, remove it from here and add this file in .gitignore

File: app/models/company.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  has_many :attendances
+  has_many :departments
+  has_many :projects_departments
+  has_many :comments
+  has_many :attachments
+  has_many :time_logs
+  has_many :events
+  has_many :company_settings
+  has_many :notifications
+  belongs_to :users, foreign_key: 'owner_id'
```
Comment (bk-az) [inline app/models/company.rb]: 1. always use singular name for one-to-one relation. `users` is not a singular name. 
2. on line # 5 users association is already defined why overwriting? 

the association name should be owner here

File: app/models/event.rb [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+# Event model
+class Event < ApplicationRecord
+  belongs_to :company
+  belongs_to :user, foreign_key: 'craeted_by_id'
```
Comment (bk-az) [inline app/models/event.rb]: typo `craeted_by_id` -> `created_by_id`
association name should be `created_by`

File: app/models/notification.rb [hunk 16]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+# Notification model
+class Notification < ApplicationRecord
```
Comment (bk-az) [inline app/models/notification.rb:4]: associations missing

File: app/models/task.rb [hunk 17]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Task model
+class Task < ApplicationRecord
+  belongs_to :company
+  belongs_to :project
+  belongs_to :user, foreign_key: 'reviewer_id'
```
Comment (bk-az) [inline app/models/task.rb]: association name should be `reviewer`

File: app/models/task.rb [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+  belongs_to :watcher, polymorphic: true
+  belongs_to :user, foreign_key: 'assigned_to_id' # confusion
```
Comment (bk-az) [inline app/models/task.rb]: association name should be assigned_to, and polymorphic

File: app/models/task.rb [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+  belongs_to :user, foreign_key: 'created_id'
```
Comment (bk-az) [inline app/models/task.rb]: `created_id`?

File: app/models/team.rb [hunk 20]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+# Team model
+class Team < ApplicationRecord
+  belongs_to :user, foreign_key: 'lead_id'
```
Comment (bk-az) [inline app/models/team.rb]: association name should be lead

File: app/models/user.rb [hunk 21]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+# Users model
+class User < ApplicationRecord
+  has_one :company
```
Comment (bk-az) [inline app/models/user.rb]: has_one :company
belongs_to :company

same name in both associations?

File: db/migrate/20200907135832_create_teams.rb [hunk 22]
```diff
@@ -0,0 +1,12 @@
+class CreateTeams < ActiveRecord::Migration[6.0]
+  def change
+    create_table :teams do |t|
+      t.string :name, null: false
+      t.text :description
+      t.integer :company_id, null: false
+      t.integer :lead_id, null: false
+      t.integer :sequence_number, null: false
```
Comment (bk-az) [inline db/migrate/20200907135832_create_teams.rb]: sequence_number -> sequence_num

File: db/migrate/20200907135905_create_tasks.rb [hunk 23]
```diff
@@ -0,0 +1,25 @@
+class CreateTasks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :tasks do |t|
+      t.integer :company_id, null: false
+      t.string :title, null: false
+      t.text :description
+      t.integer :assigned_to_id
+      t.integer :assigned_to_type, null: false
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: why integer?
why null: false?

File: db/migrate/20200907135905_create_tasks.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+      t.integer :created_by_id, null: false
+      t.string :status, default: 'New'
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: make it null: false, don't specify default value

extra note for future ref: store lower case values for status

File: db/migrate/20200907135905_create_tasks.rb [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+      t.datetime :due_date
+      t.integer :project_id, null: false
+      t.integer :reviewer_id
+      t.integer :watcher_id
+      t.integer :watcher_type
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: why using integer for type column?

File: db/migrate/20200907135905_create_tasks.rb [hunk 26]
```diff
@@ -0,0 +1,25 @@
...
+      t.datetime :started_at, null: false
+      t.datetime :completed_at
+      t.integer :progress # in percentage
+      t.datetime :expected_end_date, null: false
+      t.datetime :expected_start_date, null: false
+      t.integer :priority
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: null: false

File: db/migrate/20200907135905_create_tasks.rb [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+      t.string :status, default: 'New'
+      t.datetime :due_date
+      t.integer :project_id, null: false
+      t.integer :reviewer_id
+      t.integer :watcher_id
+      t.integer :watcher_type
+      t.datetime :started_at, null: false
+      t.datetime :completed_at
+      t.integer :progress # in percentage
+      t.datetime :expected_end_date, null: false
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: remove null: false from `expected_end_date` and `expected_start_date`

File: db/migrate/20200908065920_create_projects.rb [hunk 28]
```diff
@@ -0,0 +1,16 @@
+class CreateProjects < ActiveRecord::Migration[6.0]
+  def change
+    create_table :projects do |t|
+      t.string :name, null: false
+      t.integer :company_id, null: false
+      t.text :description
+      t.string :status
```
Comment (bk-az) [inline db/migrate/20200908065920_create_projects.rb]: null: false

File: db/migrate/20200908065920_create_projects.rb [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+      t.datetime :start_date, null: false
+      t.datetime :end_date, null: false
+      t.datetime :expected_start_date, null: false
+      t.datetime :expected_end_date, null: false
```
Comment (bk-az) [inline db/migrate/20200908065920_create_projects.rb]: remove null: false from all of these dates

File: db/migrate/20200907135905_create_tasks.rb [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+      t.integer :assigned_to_id
+      t.integer :assigned_to_type, null: false
+      t.integer :created_by_id, null: false
+      t.string :status, default: 'New'
+      t.datetime :due_date
+      t.integer :project_id, null: false
+      t.integer :reviewer_id
+      t.integer :watcher_id
+      t.integer :watcher_type
+      t.datetime :started_at, null: false
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: remove null: false

File: db/migrate/20200907135905_create_tasks.rb [hunk 31]
```diff
@@ -0,0 +1,25 @@
...
+      t.datetime :completed_at
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: started_at -> start_date
completed_at -> end_date

File: db/migrate/20200908065920_create_projects.rb (hunk 29 above)
Comment (bk-az) [reply db/migrate/20200908065920_create_projects.rb]: t.datetime :start_date, null: false
      t.datetime :end_date, null: false
      t.datetime :expected_start_date, null: false
      t.datetime :expected_end_date, null: false

File: db/migrate/20200908070038_create_attendances.rb [hunk 32]
```diff
@@ -0,0 +1,12 @@
+class CreateAttendances < ActiveRecord::Migration[6.0]
+  def change
+    create_table :attendances do |t|
+      t.integer :company_id, null: false
+      t.integer :user_id, null: false
+      t.datetime :login_time
+      t.datetime :logout_time
+      t.boolean :present, default: true
```
Comment (bk-az) [inline db/migrate/20200908070038_create_attendances.rb]: null: false

File: db/migrate/20200908070148_create_comments.rb [hunk 33]
```diff
@@ -0,0 +1,11 @@
+class CreateComments < ActiveRecord::Migration[6.0]
+  def change
+    create_table :comments do |t|
+      t.integer :company_id, null: false
+      t.integer :commentable_id, null: false
+      t.integer :commentable_type, null: false
```
Comment (bk-az) [inline db/migrate/20200908070148_create_comments.rb]: integer?

File: db/migrate/20200908070243_create_time_logs.rb [hunk 34]
```diff
@@ -0,0 +1,11 @@
+class CreateTimeLogs < ActiveRecord::Migration[6.0]
+  def change
+    create_table :time_logs do |t|
+      t.integer :task_id, null: false
```
Comment (bk-az) [inline db/migrate/20200908070243_create_time_logs.rb]: remove null: false, so we can provide option to log time globally

File: db/migrate/20200908070243_create_time_logs.rb [hunk 35]
```diff
@@ -0,0 +1,11 @@
...
+      t.integer :company_id, null: false
+      t.integer :user_id, null: false
+      t.decimal :hours, precision: 5, scale: 3
```
Comment (bk-az) [inline db/migrate/20200908070243_create_time_logs.rb]: scale: 2

File: db/migrate/20200908070327_create_notifications.rb [hunk 36]
```diff
@@ -0,0 +1,11 @@
+class CreateNotifications < ActiveRecord::Migration[6.0]
+  def change
+    create_table :notifications do |t|
+      t.integer :company_id, null: false
+      t.string :title, null: false
+      t.text :description
+      t.boolean :sent		# confusion
```
Comment (bk-az) [inline db/migrate/20200908070327_create_notifications.rb]: rename to `read`, default false, null false

File: db/migrate/20200907135958_create_users_teams.rb [hunk 37]
```diff
@@ -0,0 +1,10 @@
+class CreateUsersTeams < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users_teams do |t|
+      t.integer :team_id, null: false
```
Comment (bk-az) [inline db/migrate/20200907135958_create_users_teams.rb]: pass additional option `index: true` to all foreign keys, for polymorphic case add a composite index on `xyz_id` and `xyz_type` columns,

File: app/models/project.rb [hunk 38]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+# Project model
+class Project < ApplicationRecord
+  belongs_to :comapny_id
+  belongs_to :user, forein_key: 'created_by_id'
```
Comment (bk-az) [inline app/models/project.rb]: created_by

File: app/models/task.rb (hunk 19 above)
Comment (bk-az) [inline app/models/task.rb]: association name -> created_by

File: app/models/team.rb [hunk 39]
```diff
@@ -0,0 +1,10 @@
...
+  belongs_to :company
+  has_many :users_team
+  has_many :users, through: :users_team
+  has_many :tasks, as: :watcher
```
Comment (bk-az) [inline app/models/team.rb]: association name -> `watching_tasks`
missing association `assigned_tasks`

File: db/migrate/20200907135905_create_tasks.rb [hunk 40]
```diff
@@ -0,0 +1,25 @@
+class CreateTasks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :tasks do |t|
+      t.integer :company_id, null: false
+      t.string :title, null: false
+      t.text :description
+      t.integer :assigned_to_id
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb]: assigned_to -> assignable

File: app/models/user.rb (hunk 21 above)
Comment (bk-az) [inline app/models/user.rb]: association name: `owned_company` or any suitable name

File: app/models/user.rb [hunk 41]
```diff
@@ -0,0 +1,17 @@
...
+  belongs_to :company
+  belongs_to :department
+  has_one :team
```
Comment (bk-az) [inline app/models/user.rb]: why has_one?

File: app/models/user.rb [hunk 42]
```diff
@@ -0,0 +1,17 @@
...
+  has_many :users_team
+  has_many :tasks, through: :users_team
```
Comment (bk-az) [inline app/models/user.rb]: correct this, 

before sharing the link in the group for review, always review it yourself and add a comment 'cycle 0'

File: app/models/user.rb [hunk 43]
```diff
@@ -0,0 +1,17 @@
...
+  has_many :tasks, as: :watcher
```
Comment (bk-az) [inline app/models/user.rb]: watching_tasks

missing association: assigned_tasks

File: app/models/user.rb [hunk 44]
```diff
@@ -0,0 +1,17 @@
...
+  has_many :projects
```
Comment (bk-az) [inline app/models/user.rb]: how?

Comment (bk-az) [reply app/models/user.rb]: should be `created_projects`, specify proper options

File: app/models/user.rb [hunk 45]
```diff
@@ -0,0 +1,17 @@
...
+  has_many :attendances
+  has_many :time_logs
+  has_many :events
```
Comment (bk-az) [inline app/models/user.rb]: created_events

File: app/models/user.rb (hunk 43 above)
Comment (bk-az) [reply app/models/user.rb]: missing association: created_tasks

File: db/migrate/20200907135535_create_users.rb [hunk 46]
```diff
@@ -0,0 +1,15 @@
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
+      t.string :first_name, null: false
+      t.string :last_name, null: false
+      t.string :email, null: false, uniqueness: true
+      t.integer :role_id, null: false
```
Comment (bk-az) [inline db/migrate/20200907135535_create_users.rb]: add index on role_id as well

File: app/models/user.rb (hunk 43 above)
Comment (bk-az) [reply app/models/user.rb]: missing association: created_teams (add column in teams)

Comment (bk-az) [reply app/models/user.rb]: missing association: leading_teams (inverse of `lead` in Team), any suitable name can be used instead of leading_teams

Comment (bk-az) [reply app/models/user.rb]: missing association: reviewing_tasks

File: app/models/comment.rb [hunk 47]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+# Comment model
+class Comment < ApplicationRecord
+  belongs_to :company
+  belongs_to :user
+  belongs_to :commentable, polymorphic: true
+  has_many :attachments, as: :attachable
```
Comment (bk-az) [inline app/models/comment.rb]: `dependent: destroy`

File: app/models/company.rb [hunk 48]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+# Company model
+class Company < ApplicationRecord
```
Comment (bk-az) [inline app/models/company.rb:4]: status column

File: app/models/department.rb [hunk 49]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+# Department model
+class Department < ApplicationRecord
+  has_many :users
+  has_many :projects_departments
```
Comment (bk-az) [inline app/models/department.rb]: `dependent: destroy`

File: app/models/department.rb [hunk 50]
```diff
@@ -0,0 +1,9 @@
...
+  has_many :projects, through: :projects_departments
```
Comment (bk-az) [inline app/models/department.rb:7]: skip dependent destroy

File: app/models/project.rb [hunk 51]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+# Project model
+class Project < ApplicationRecord
+  belongs_to :company_id
```
Comment (bk-az) [inline app/models/project.rb]: belongs_to :company

File: app/models/task.rb [hunk 52]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+# Task model
+class Task < ApplicationRecord
+  belongs_to :company
+  #belongs_to :project
+  belongs_to :project, foreign_key: 'parent_id'
```
Comment (bk-az) [inline app/models/task.rb]: remove

File: app/models/task.rb [hunk 53]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+# Task model
+class Task < ApplicationRecord
+  belongs_to :company
+  #belongs_to :project
```
Comment (bk-az) [inline app/models/task.rb]: uncomment

File: app/models/task.rb [hunk 54]
```diff
@@ -0,0 +1,18 @@
...
+  belongs_to :project, foreign_key: 'parent_id'
+  belongs_to :reviewer, class_name: 'User'
+  belongs_to :watcher, polymorphic: true
```
Comment (bk-az) [inline app/models/task.rb]: remove

File: app/models/task.rb [hunk 55]
```diff
@@ -0,0 +1,18 @@
...
+  belongs_to :assignable, polymorphic: true
+  belongs_to :created_by, class_name: 'User'
+  has_many :comments, as: :commentable
+  has_many :time_logs
+  has_many :attachments, as: :attachable
+  has_many :tasks_watchers
+  has_many :watchers, through: :tasks_watchers, class_name: 'TasksWatcher'
```
Comment (bk-az) [inline app/models/task.rb]: in has many through case, the class_name, foreign_key and primary_key are ignored

Comment (bk-az) [inline app/models/task.rb]: `has_many :watcher_users, through: :tasks_watchers, source: watcher, source_type: 'User'`
`has_many :watcher_teams, through: :tasks_watchers, source: watcher, source_type: 'Team'`

File: app/models/task.rb [hunk 56]
```diff
@@ -0,0 +1,18 @@
...
+  belongs_to :parent, class_name: 'Task'
```
Comment (bk-az) [inline app/models/task.rb:17]: `has_many :childs, foreign_key: 'parent_id', class_name: 'Task'`

File: app/models/team.rb [hunk 57]
```diff
@@ -0,0 +1,12 @@
...
+
+# Team model
+class Team < ApplicationRecord
+  belongs_to :lead, class_name: 'User'
+  belongs_to :company
+  has_many :users_team
+  has_many :users, through: :users_team
+  has_many :assigned_tasks, class_name: 'Task', as: :assignable
+  has_many :tasks_watchers, as: :watcher
+  has_many :tasks, through: :tasks_watchers
```
Comment (bk-az) [inline app/models/team.rb]: has_many :watching_tasks, through: :tasks_watchers, source: task

File: app/models/user.rb [hunk 58]
```diff
@@ -0,0 +1,27 @@
...
+# Users model
+class User < ApplicationRecord
+  has_one :owned_company, class_name: 'Company'
+  belongs_to :company
+  belongs_to :department
+  belongs_to :team
+  has_many :users_team
+  has_many :comments
+  has_many :tasks
+  has_many :created_projects, class_name: 'Project'
```
Comment (bk-az) [inline app/models/user.rb]: foreign_key: 'created_by_id'

File: app/models/user.rb [hunk 59]
```diff
@@ -0,0 +1,27 @@
...
+  has_many :attendances
+  has_many :time_logs
+  has_many :created_events, class_name: 'Event'
+  has_many :attachments, as: :attachable
+  has_many :tasks, as: :assignable
```
Comment (bk-az) [inline app/models/user.rb]: assigned_tasks

File: app/models/user.rb [hunk 60]
```diff
@@ -0,0 +1,27 @@
...
+  has_many :assigned_tasks, class_name: 'Task'
```
Comment (bk-az) [inline app/models/user.rb]: remove

File: app/models/user.rb [hunk 61]
```diff
@@ -0,0 +1,27 @@
...
+  has_many :created_tasks, class_name: 'Task'
+  has_many :created_teams, class_name: 'Team'
+  has_many :leading_teams, class_name: 'Team'
+  has_many :reviewing_tasks, class_name: 'Task'
+  has_many :tasks_watchers, as: :watcher
+  has_many :tasks, through: :tasks_watchers
```
Comment (bk-az) [inline app/models/user.rb]: watching_tasks, source

File: db/migrate/20200907135535_create_users.rb [hunk 62]
```diff
@@ -0,0 +1,18 @@
...
+# User Migration
+class CreateUsers < ActiveRecord::Migration[6.0]
+  def change
+    create_table :users do |t|
+      t.string :first_name, null: false
+      t.string :last_name, null: false
+      t.string :email, null: false, uniqueness: true
+      t.integer :role_id, null: false, index: true
+      t.integer :company_id, null: false, index: true
+      t.integer :sequence_number, null: false
```
Comment (bk-az) [inline db/migrate/20200907135535_create_users.rb]: index: true

Comment (bk-az) [reply db/migrate/20200907135535_create_users.rb]: ctrl + shift + f, global search for sequence_number

File: db/migrate/20200907135905_create_tasks.rb [hunk 63]
```diff
@@ -0,0 +1,27 @@
...
+
+# Task Migration
+class CreateTasks < ActiveRecord::Migration[6.0]
+  def change
+    create_table :tasks do |t|
+      t.integer :company_id, null: false
+      t.string :title, null: false
+      t.text :description
+      t.integer :assignable_id
+      t.string :assignable_type
```
Comment (bk-az) [inline db/migrate/20200907135905_create_tasks.rb:11]: add_index(assignable)

File: db/migrate/20200908065920_create_projects.rb [hunk 64]
```diff
@@ -0,0 +1,19 @@
...
+    create_table :projects do |t|
+      t.string :name, null: false
+      t.integer :company_id, null: false, index: true
+      t.text :description
+      t.string :status, null: false
+      t.datetime :start_date
+      t.datetime :end_date
+      t.datetime :expected_start_date
+      t.datetime :expected_end_date
+      t.datetime :created_by_id, null: false
```
Comment (bk-az) [inline db/migrate/20200908065920_create_projects.rb]: index: true

File: db/migrate/20200908070148_create_comments.rb [hunk 65]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+# Comment Migration
+class CreateComments < ActiveRecord::Migration[6.0]
+  def change
+    create_table :comments do |t|
+      t.integer :company_id, null: false, index: true
+      t.integer :commentable_id, null: false
+      t.string :commentable_type, null: false
+      add_index(:comments, %i[commentable_id commentable_type])
```
Comment (bk-az) [inline db/migrate/20200908070148_create_comments.rb]: move outside to create_table

File: test/fixtures/users_teams.yml [hunk 66]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (bk-az) [inline test/fixtures/users_teams.yml]: delete test/fixtures and test/models

File: app/models/task.rb [hunk 67]
```diff
@@ -3,16 +3,15 @@
...
-  has_many :tasks_watchers
-  has_many :watchers, through: :tasks_watchers, class_name: 'TasksWatcher'
-  belongs_to :parent, class_name: 'Task'
+  has_many :comments, as: :commentable, dependent: :destroy
+  has_many :time_logs, dependent: :destroy
+  has_many :attachments, as: :attachable, dependent: :destroy
+  has_many :tasks_watchers, dependent: :destroy
+  has_many :watcher_users, through: :tasks_watchers, source: :watcher, source_type: 'User'
+  has_many :watcher_teams, through: :tasks_watchers, source: :watcher, source_type: 'Team'
+  has_many :childs, foreign_key: 'parent_id', class_name: 'Task'
```
Comment (bk-az) [inline app/models/task.rb:16]: add parent association as well
belongs_to :parent, class_name: 'Task'

File: app/models/user.rb [hunk 68]
```diff
@@ -6,22 +6,21 @@ class User < ApplicationRecord
...
-  has_many :created_tasks, class_name: 'Task'
-  has_many :created_teams, class_name: 'Team'
-  has_many :leading_teams, class_name: 'Team'
-  has_many :reviewing_tasks, class_name: 'Task'
-  has_many :tasks_watchers, as: :watcher
-  has_many :tasks, through: :tasks_watchers
-  has_many :projects_users
+  has_many :users_team, dependent: :destroy
+  has_many :comments, dependent: :destroy
+  has_many :tasks, dependent: :destroy
```
Comment (bk-az) [inline app/models/user.rb]: remove

File: app/models/user.rb [hunk 69]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+# Users model
+class User < ApplicationRecord
+  has_one :owned_company, class_name: 'Company'
```
Comment (abdul674) [inline app/models/user.rb]: remove this,

File: config/database.yml [hunk 70]
```diff
@@ -14,7 +14,7 @@ default: &default
   encoding: utf8mb4
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   username: root
-  password:
+  password: password
   socket: /var/run/mysqld/mysqld.sock
 
 development:
```
Comment (abdul674) [inline config/database.yml]: remove this file from git.

**Pull request merged**